### PR TITLE
Packaging as namespace packages

### PR DESCRIPTION
**Pull request by untereiner**

### CODE REVIEW

**Comment by untereiner** [comment]
@cssherman I have finished the work for this PR

**Comment by cssherman** [comment]
> @cssherman I have finished the work for this PR Also:
> 
>     * would it be possible to rename `geosPythonPackages` into `geos-python` ?
> 
>     * afterwards I would like to publish this packages to pypi

I'll start taking a look!  I don't see any issues with the name change

File: docs/conf.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 # Add python modules to be documented
 python_root = '..'
-python_modules = ( 'geosx_mesh_tools_package', 'geosx_xml_tools_package', 'geosx_mesh_doctor', 'geos_ats_package',
+python_modules = ( 'geosx_mesh_tools_package', 'geosx_xml_tools_package', 'geosx_mesh_doctor', 'geos-ats_package',
```
Comment (cssherman) [inline docs/conf.py:21]: The package names, paths that are used in the python autodocs need to be updated.  We can easily add a CI test to make sure docs build correctly before deployment.

File: geos-timehistory/src/geos/timehistory/plot.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import numpy as np
-from hdf5_wrapper import hdf5_wrapper as h5w
+from geos.hdf5.wrapper import hdf5_wrapper as h5w
```
Comment (cssherman) [inline geos-timehistory/src/geos/timehistory/plot.py:2]: Given the name changes, shouldn't this be:

`from geos.hdf5wrapper import wrapper as h5w`

**Comment by cssherman** [comment]
Since we are currently only testing package installation in the CI, we need to do a bunch of manual tests to make sure that the changes work.  Also, @untereiner - could you open up a testing branch in the main GEOS repository, following these instructions:

https://geosx-geosx.readthedocs-hosted.com/projects/geosx-geospythonpackages/en/latest/#development-debugging

**Comment by cssherman** [comment]
>     * afterwards I would like to publish this packages to pypi

Before we do this, we need to clarify the license information in these packages / the top-level repo.  We also need to check with @rrsettgast and others to get approval to do so.

**Comment by cssherman** [comment]
@untereiner - I'm also happy to setup the pip repository.  I'm already managing one for another project, so I'm comfortable with the tools/workflows.

**Comment by untereiner** [comment]
> Since we are currently only testing package installation in the CI, we need to do a bunch of manual tests to make sure that the changes work. Also, @untereiner - could you open up a testing branch in the main GEOS repository, following these instructions:
> 
> https://geosx-geosx.readthedocs-hosted.com/projects/geosx-geospythonpackages/en/latest/#development-debugging

The CI fails in the GEOS repo for this branch because I changed the names of the modules, Then, the integrated tests don't find the geos-ats module.

**Comment by untereiner** [comment]
thank you @cssherman

**Comment by untereiner** [comment]
@cssherman what do you think ? I think https://github.com/GEOS-DEV/GEOS/pull/3057 is ok

**Comment by cssherman** [comment]
> @cssherman what do you think ? I think [GEOS-DEV/GEOS#3057](https://github.com/GEOS-DEV/GEOS/pull/3057) is ok

It's been a few weeks since I've looked at this, but I think that I had nailed a few of the errors that have popped up.  IIRC, the changes here (specifically moving to a pure toml config) lead to pip failures.

**Comment by untereiner** [comment]
What are the failures ? pyproject.toml is standard

**Comment by cssherman** [comment]
> What are the failures ? pyproject.toml is standard

Yes, but this is a pretty new standard.  Using a version of pip that was only 2 versions old, this lead to silent errors that were very hard to debug.  I don't think there is a way to enforce this, so we need to add some big warning boxes to the docs.

**Comment by untereiner** [comment]
When you say two versions old, you start with python 3.12 ?
Each version of python is distributed with the last version (at that moment) of pip
Regarding the status of versions of python it should be a moving window (not too large) of allowed versions.
For example the year 2024 -> [3.9, 3.12]. then for 2015 -> [3.10, 3.13], etc.
This will enforce the min pip version allowed to build the project.
This can be stated in the documentation

**Comment by cssherman** [comment]
@untereiner - do we have regular users of mesh doctor that we can call on to test these changes?  We don't have any integrated tests involving that package in the main repo (yet).

**Comment by cssherman** [comment]
@untereiner - Looking at these changes again, there were a handful of module name changes that added more depth that I like in the namespace.  There were also a number of instances where the changes weren't reflected properly in test scripts, documentation, etc.  I've done some additional work to address this in my recent commit.

**Comment by cssherman** [comment]
I've also fixed some doc build issues, and consolidated some of the older mesh conversion tools and the doctor into a single 'geos-mesh' package

**Comment by cssherman** [comment]
> When you say two versions old, you start with python 3.12 ? Each version of python is distributed with the last version (at that moment) of pip Regarding the status of versions of python it should be a moving window (not too large) of allowed versions. For example the year 2024 -> [3.9, 3.12]. then for 2015 -> [3.10, 3.13], etc. This will enforce the min pip version allowed to build the project. This can be stated in the documentation



> When you say two versions old, you start with python 3.12 ? Each version of python is distributed with the last version (at that moment) of pip Regarding the status of versions of python it should be a moving window (not too large) of allowed versions. For example the year 2024 -> [3.9, 3.12]. then for 2015 -> [3.10, 3.13], etc. This will enforce the min pip version allowed to build the project. This can be stated in the documentation


I meant two versions old of pip (obviously not an ideal situation, but this was on a CI machine that may not have been updated in a while).  While `pyproject.toml` files have been standard for a while, pure-pyproject.toml projects (i.e.: no setup.cfg) only seem to work with very recent versions of pip.

And with respect to documentation, I agree that we should note this behavior.  For most users though, we just need to include a `python -m pip install --upgrade` step in the instructions (most users are still going to use pip, regardless of our support for other tools).

For python version support, I think that we need to support all non-deprecated versions of python.  Some end-users may not have full control of their environments and we should do our best to make their lives easier.

**Comment by cssherman** [comment]
@CusiniM - do you have any thoughts about the re-organization of these packages?  The big change here is how they would be imported.  For example, `from geos_ats.test_builder import TestDeck` would now be `from geos.ats.test_builder import TestDeck`

**Comment by CusiniM** [comment]
> @untereiner - do we have regular users of mesh doctor that we can call on to test these changes? We don't have any integrated tests involving that package in the main repo (yet).

@jhuang2601 and I can test this. However the installation was already not working.

**Comment by CusiniM** [comment]
> @CusiniM - do you have any thoughts about the re-organization of these packages? The big change here is how they would be imported. For example, `from geos_ats.test_builder import TestDeck` would now be `from geos.ats.test_builder import TestDeck`

I don't think this is an issue. However, is there a specific reason for creating all those nested folders?

**Comment by cssherman** [comment]
> > @CusiniM - do you have any thoughts about the re-organization of these packages? The big change here is how they would be imported. For example, `from geos_ats.test_builder import TestDeck` would now be `from geos.ats.test_builder import TestDeck`
> 
> I don't think this is an issue. However, is there a specific reason for creating all those nested folders? almost feels like we could ship everything as a single package at this point with a single `toml`?

I haven't packaged something like this before, but from what I understand, this allows us to independently package components of a single larger module.   Maybe this has benefits for simplifying testing?

**Comment by untereiner** [comment]
They can be packaged, tested, distributed separately but they can work together easily because they share the same root. It’s a kind of separation of concerns.

**Comment by untereiner** [comment]
@cssherman  For [python versions](https://devguide.python.org/versions/) it means min 3.8 until 2024/10
Also, python is very dynamic (in many ways) then sticking to old versions seems to be a bad idea for me even if updating CI machines is painful. 
@CusiniM Each package can have it's own set of dependencies. There aren't necessarily cross dependencies.
Then:
- as a final user wanting to preprocess my meshes I can do `pip install geos-mesh`
- in the CI it is a `pip install geos-ats`.

What can I do now to help move forward ?
For the packaging, the next would be to update the CI to publish the package to pypi (github doesn't have a python package registry)..

**Comment by cssherman** [comment]
> @cssherman For [python versions](https://devguide.python.org/versions/) it means min 3.8 until 2024/10 Also, python is very dynamic (in many ways) then sticking to old versions seems to be a bad idea for me even if updating CI machines is painful. @CusiniM Each package can have it's own set of dependencies. There aren't necessarily cross dependencies. Then:
> 
>     * as a final user wanting to preprocess my meshes I can do `pip install geos-mesh`
> 
>     * in the CI it is a `pip install geos-ats`.
> 
> 
> What can I do now to help move forward ? For the packaging, the next would be to update the CI to publish the package to pypi (github doesn't have a python package registry)..

I think that this is ready to move forward.  I've added the associated PR in the main GEOS repo to the dashboard, with a note on the procedure to handle the merge.  We also need to have a discussion with the larger GEOS community to give them an explanation of the changes, so they aren't caught off-guard.

And WRT python versions, I agree that in a perfect world, we should only need to target reasonably up-to-date versions of python.  At least in short-term, I think we need to keep with the current strategy.

**Comment by untereiner** [comment]
@cssherman great!
It will be the right time to also rename the repository from `geosPythonPackages` into a more pythonic style like `geos-python`.

**Review by cssherman — APPROVED** [review]

**Pull request merged**